crates/services/gas_price_service/src/v1/da_source_adapter/block_committer_costs.rs

```diff
@@ -6,57 +9,334 @@ use crate::v1::da_source_adapter::{
 };
 
-/// This struct is used to denote the block committer da gas price source,,
+#[async_trait::async_trait]
+trait BlockCommitterApi: Send + Sync {
+    // Used on first run to get the latest costs and seqno
+    async fn get_latest_costs(&self) -> DaBlockCostsResult<Option<RawDaBlockCosts>>;
+    // Used to get the costs for a specific seqno
+    async fn get_costs_by_seqno(
+        &self,
+        number: u64,
+    ) -> DaBlockCostsResult<Option<RawDaBlockCosts>>;
+    async fn get_bundles_by_range(
+        &self,
+        range: core::ops::Range<u64>,
+    ) -> DaBlockCostsResult<Vec<Option<RawDaBlockCosts>>>;
+}
+
+/// This struct is used to denote the block committer da block costs source
 /// which receives data from the block committer (only http api for now)
-pub struct BlockCommitterDaBlockCosts {
-    client: reqwest::Client,
-    url: Url,
+pub struct BlockCommitterDaBlockCosts<BlockCommitter> {
+    client: BlockCommitter,
+    last_value: Option<RawDaBlockCosts>,
 }
 
 #[derive(Debug, Deserialize, Serialize, Clone, Default, PartialEq)]
-struct RawDaBlockCosts {
-    pub l2_block_range: core::ops::Range<u32>,
-    pub blob_size_bytes: u32,
-    pub blob_cost: u128,
+pub struct RawDaBlockCosts {
+    // Sequence number (Monotonically increasing nonce)
+    pub sequence_number: u64,
+    // The range of blocks that the costs apply to
+    pub blocks_range: core::ops::Range<u64>,
+    // The DA block height of the last transaction for the range of blocks
+    pub da_block_height: DaBlockHeight,
+    // Rolling sum cost of posting blobs (wei)
+    pub total_cost: u128,
+    // Rolling sum size of blobs (bytes)
+    pub total_size_bytes: u32,
+}
+
+impl<BlockCommitter> BlockCommitterDaBlockCosts<BlockCommitter> {
+    /// Create a new instance of the block committer da block costs source
+    pub fn new(client: BlockCommitter, last_value: Option<RawDaBlockCosts>) -> Self {
+        Self { client, last_value }
+    }
 }
 
-impl From<RawDaBlockCosts> for DaBlockCosts {
-    fn from(raw: RawDaBlockCosts) -> Self {
-        DaBlockCosts {
-            l2_block_range: raw.l2_block_range,
-            blob_size_bytes: raw.blob_size_bytes,
-            blob_cost_wei: raw.blob_cost,
+#[async_trait::async_trait]
+impl<BlockCommitter> DaBlockCostsSource for BlockCommitterDaBlockCosts<BlockCommitter>
+where
+    BlockCommitter: BlockCommitterApi,
+{
+    async fn request_da_block_cost(&mut self) -> DaBlockCostsResult<DaBlockCosts> {
+        let response;
+
+        if let Some(last_value) = &self.last_value {
+            response = self
+                .client
+                .get_costs_by_seqno(last_value.sequence_number + 1)
+                .await?;
+        } else {
+            // we have to error if we cannot find the first set of costs
+            response = self.client.get_latest_costs().await?;
+        }
+
+        if let Some(response) = response {
+            let res;
+            if let Some(last_value) = &self.last_value {
```

Comment:
nit: I think I prefer to only have one `if let Some(last_value) = &self.last_value` and have redundant `if let Some(response) = response` in each branch.

Reply:
addressed in https://github.com/FuelLabs/fuel-core/commit/edfc1ffc344ebaf603663cab679c6007d7c0634a and https://github.com/FuelLabs/fuel-core/commit/28932883d0086d342e68e2a212bdce93796ad2c1